cluster-api/pkg/controller/machineset/controller.go

```diff
@@ -79,6 +82,17 @@ func (c *MachineSetControllerImpl) syncReplicas(machineSet *v1alpha1.MachineSet,
 	desiredReplicas := *machineSet.Spec.Replicas
 	diff := int(currentMachineCount - desiredReplicas)
 
+	// Take ownership of machines if not already owned.
+	for _, machine := range machines {
+		if !metav1.IsControlledBy(machine, machineSet) {
+			glog.Infof("MachineSet (%v) does not own machine (%v).", machineSet.Name, machine.Name)
+			machine.ObjectMeta.OwnerReferences = []metav1.OwnerReference{*metav1.NewControllerRef(machineSet, controllerKind),}
```

Comment:
I think the situation is a little more nuanced than this. For each machine you should:

- See if it already has a ControllerRef
- If it does
  - If it's not this machine set, leave it alone
  - If it is this machine set
    - If the machine set is being deleted, leave it alone
    - If the labels no longer match, delete your owner ref and orphan it (though you selected this by labels so this logic might be elsewhere)
- If it does not have a ControllerRef
  - If you're being deleted, do nothing
  - If the labels don't match, do nothing
  - If the labels match, **ADD** an owner reference for yourself. Don't replace wholesale.

Reply:
Made some changes to handle the scenarios you outlined. 
The reconciler seems to already filter out machines that doesn't match the machineset selector, so those machines aren't passed through.